### ISSUE TITLE
PDS3 can have "local" times.

### ISSUE THREAD

**Issue by rbeyer** [planetarypy/pvl#80]
**Describe the bug**
TLDR:
* Right now, `pvl.load()` always returns an "aware" `datetime` whenever *any* of its decoder classes encounters a datetime or time in the input PVL.
* Re-reading the documentation indicates that the following should be true:
   * The strict PVL decoder should always return "aware" `datetime` objects.
   * The strict ODL decoder should return "aware" and "naive" `datetime` objects since "local" times without a timezone are allowed under ODL.
* Finally, there is some uncertainty about how to treat PDS3 PVL.  Currently, there is no PDS3 decoder, we just use the ODL decoder (because I didn't think there was a difference before), but in strict PDS3, no timezone specifier is allowed, so does that mean PDS3 `datetime` objects are always "naive" because there is no specifier or always "aware" because they should always be assumed to be UTC?

So, remember the long discussion we had last spring about Python `datetime` objects(#57 ) and whether they were "naive" (no sense of the time zone they belonged to) or "aware" (they knew their offset from UTC)?  That discussion was in the context of the PVL specification, which absolutely requires that all times be interpreted as relative to UTC.  On the plus side, that's still true.

However, the ODL/PDS3 spec ... differs.  ODL says 
> A time that is not followed by either the Zulu indicator or a time zone offset is assumed to be a local time.  

In this context a "local time" would be a "naive" `datetime` object.  So we need to adjust `pvl.decoder.ODLDecoder()` so that it will return "naive" and "aware" `datetime` objects as appropriate.

Then the PDS3 specification about times says:
> Alternate time zones (e.g., YYYY-MM-DDTHH:MM:SS.SSS + HH:MM) may not be used in a PDS label. The only allowed time formats are (1) YYYY-MM-DDTHH:MM:SS.SSS. (2) YYYY-DDDTHH:MM:SS.SSS.

Note, this doesn't allow a trailing "Z" like ODL does, so we need to fix that.

I guess I had interpreted that to mean that PVL only allowed "aware", and mis-interpreted that to mean that ODL and PDS3 followed suit.  But that's not what that says.  It says that ODL allows "naive" times, and then only that PDS3 doesn't allow you to specify an offset with "+HH:MM" or "Z".  Since I had thought that ODL was like PVL, I had made the mental leap and thought that PDS3 was only UTC, but you could also mentally load the "ODL" rule, and then read the "PDS3" rule to mean that a strict PDS3 encoder should only return "naive" `datetime`s since ODL allows it, and you can't specify a time offset.

So, what to do about PDS3?  On the one hand, since the ODL spec allows "local times" and the PDS3 part doesn't contradict that then maybe that means that PDS3 "Time" values should be interpreted as "naive" `datetime` objects, and since "correct" PDS3 labels will *never* have a "+HH:MM" or "Z" suffix, you would always return "naive" `datetime` objects if parsing.  On the other hand, since PDS3 "Time" values can never specify their timezone, does that mean they should always be interpreted as UTC, and thus should always be "aware" as they are now (via the `pvl.decoder.ODLDecoder()`)?

Not sure.  Thoughts?  Help?

**To Reproduce**

ODL decoders always return "aware" `datetime` objects (`tzinfo is not None`), even when there is no offset specifier:
```python
>>> import pvl
>>> odl = pvl.loads("time = 2000-02-25T10:54:12.129", decoder=pvl.decoder.ODLDecoder())
>>> print(odl)
PVLModule([
  ('time',
   datetime.datetime(2000, 2, 25, 10, 54, 12, 129000, tzinfo=datetime.timezone.utc))
])
```

And then, the PDS3Encoder mistakenly adds a "Z" (assume `odl` from above):
```python
>>> print(pvl.dumps(odl, encoder=pvl.encoder.PDSLabelEncoder()))
TIME = 2000-02-25T10:54:12.129000Z
END
```

**Expected behavior**
- The dict-like returned by the loaders when a `pvl.decoder.ODLDecoder()` is specified, should return an "aware" `datetime` when a "+HH:MM" or a "Z" is present, but a "naive" `datetime` otherwise.
- The PVL-text written by `pvl.encoder.PDS3Encoder()` should never append a "Z" to the time.

**Your Environment (please complete the following information):**
 - OS: macOS, but should affect all OSes.
 - Environment information: Python 3.6, but should affect all versions.
 - `pvl` Version 1.1.0

**Comment by rbeyer** [comment]
I spent some time writing some code to address the above, and then realized that I may not have read the spec correctly, and that some of the above was incorrect.  I should have spent time with the specifications and *then* written code, and now regret my life choices.  Anyway, what follows is my "scholarship" on the PVL specification documents regarding time, and what that means for how we should handle them in the `pvl` library.  Apologies, this is mostly to keep my head clear (hopefully) as I go write the code.

The upshot is that the first "Expected Behavior" item is still correct, the second isn't, and there are others. Here's the roundup of changes to `pvl` 1.1.0 that need to happen to properly implement all of these time specifications (assuming I read them correctly this time):
* No changes to the PVL handling is needed (yay).
* These are the changes to ODL handling needed (sure, nobody uses ODL without going on to PDS3, but still):
    * ODLDecoder: Raise an exception if there is a value of 60 for seconds.
    * ODLDecoder: If there is no timezone specifier, return a "naive" `time` object.
    * ODLEncoder: "naive" `time` objects should raise an Exception.
* These are the changes to PVL handling needed:
    * decoding: all times returned will be "aware" and a time without a specifier is assumed to be UTC (so this is like the current PVL implementation but not what the ODL implementation *should be*).
    * decoding: +HH:MM timezone offsets should raise an Exception (so again, like PVL but not like current ODL).
    * decoding: times with precision greater than milisecond should raise an Exception.
    * encoding: `time` objects with precision greater than milisecond should raise an Exception.
* And, of course, make sure the OmniDecoder can still parse all the things.

That should be it for the changes, but to arrive at them, I had to work through the very long text below.  I'll just be using the term `time` to refer to `datetime.time`, and since the `datetime.datetime` class descends from the `datetime.time` class, all of this applies to them equally.


## PVL Spec (CCSDS Blue Book)

Section 2.3.2.1.3: 
 - "all time is represented in Universal Coordinated Time"
 - "all of this can be optionally followed by the character Z as a terminator"
 - Seconds values from 00 to 60 are allowed (60 to accommodate leap seconds).
 - More than microsecond precision is exhibited in example.

This logcial set of statements, nicely contained in a single section allows a straightforward implementation:

### Decoding, `pvl.load()`

 - PVL-text Times with a timezone offset (e.g. +HH:MM) should raise an Exception and not be parsed.
 - PVL-text Times with a trailing "Z" should also be handled.
 - 60-second time values are not representable as `time` objects, so PVL-text Times will either be parsed as `time` objects when they can be, and must just be returned as `str` when they have a value of 60 in the seconds value.
 - PVL-text Times can have more precision that is representable in a `time`, if this is the case, the PVL-text Time will be returned as a `str` instead of a `time`.
 - Any PVL-text Times that can be returned as `time` objects will be "aware" because "all PVL times are UTC."

### Encoding, `pvl.dump()`

- Allowable formats provide straightforward conversion from `time` to PVL-text.
- Since the trailing "Z" is optional, and no preferred implementaition is suggested, `pvl` developers can choose whether to place the trailing "Z" or not.  The `pvl` 1.1.0 implementation currently chooses to not do so, but could.

## ODL Spec (Chapter 12 of the PDS3 Standards Reference)

Section 12.3.2.1, Date and Time Values
- 60-second time values are *not* allowed.
- "zone offsets" are allowed (as either +/-12 hours or as +/-HH:MM)

Section 12.3.2.2, Implementation of Dates and Times
- Years must always be four-digit years
- "Times in ODL may be specified with unlimited precision," but the actual precision can be implementation specific.
- Do not output local time values, "Use either the UTC or zoned time format instead."

Section 12.3.2.5, Times
- "Times may be specified in Universal Time Coordinated (UTC) by following the time with the letter Z".
- Time zone offsets can be represented.
- "A time that is not followed by either the Zulu indicator or a time zone offset is assumed to be a local time."


Only the differences from PVL handling are specified below, and assume inheritance from PVL handling above.

### Decoding, `pvl.load()`
- PVL-text Times with a timezone offset (e.g. +HH:MM) will be parsed and returned as an "aware" `time` object.
- Since 60-second time values are not allowed, they should raise an Exception.
- Returned `time` objects will only be "aware" if there is a trailing "Z" or a timezone offset present, and "naive" otherwise.


### Encoding, `pvl.dump()`
- "aware" `time` objects should always be written with an offset (or trailing "Z" if offset is zero).
- "naive" `time` objects should not be serialized, and should raise an Exception.  So this would be a case where some PVL-text could be loaded under ODL, but the decoded Python objects could not be written back out under ODL, because ODL is asymmetric in this way.  Go figure.



## PDS3 Spec (Chapter 12 of the PDS3 Standards Reference)

Section 12.3.2.3, PDS Implementation of Dates and Times
- Software should interpret "label times as UTC times."
- "On output, a 'Z' will be appended to label times."

Section 12.7.3, ODL/PVL Usage, Sub-para 14:
> Alternate time zones (e.g., YYYY-MM-DDTHH:MM:SS.SSS + HH:MM) may not be used in a PDS label. The only allowed time formats are
> (1) YYYY-MM-DDTHH:MM:SS.SSS.
> (2) YYYY-DDDTHH:MM:SS.SSS.
>
> See Section 7.3.2(6) for a more detailed description.

The section 7.3.2(6) does not appear in this document that I could find.

If one reads this para in isolation, one might interpret this as indicating that *no* timezone specifier is allowed, however, this combined with Section 12.3.2.3 indicates that a trailing "Z" is allowed on read, and required on write.

Section 12.7.3, ODL/PVL Usage, Sub-para 14:
- Dates should always have zeroes: "2001-04-01", not "2001-4-1"


Only the differences from ODL handling are specified below, and assume inheritance from ODL and PVL handling above.

### Decoding, `pvl.load()`
- PVL-text times without a timezone specifier will be interpreted "as UTC times" and so while ODL would decode them to "naive" times, PDS should decode them to "aware" times and assume UTC.  As a consequence, all `times` returned will be "aware".
- PVL-text Times with a "+HH:MM" timezone specifier should raise an Exception since they "may not be used in a PDS label."  So this is unlike ODL, but like PVL.
- PVL-text Times with a precision better than milisecond should raise an exception.  ODL parsing allows mircrosecond precsion to be parsed into a `time` and times with additional precision to be returned as strings, but PDS should not allow either.

### Encoding, `pvl.dump()`
- "aware" `time` objects with offsets other than zero should raise an Exception.
- `time` objects with precision greater than milisecond should raise an Exception.
- `time` objects should always be written out with a trailing "Z".
- `time` objects should always be written out left-padded with zeroes.
